Proposal of simple JWK support

anakinj opened the pull request:
This PR is related to issue #158

Been using this gem for handling JWT tokens but with the lack of JWK support been forced to also include another JWT gem in the picture (https://rubygems.org/gems/json-jwt).

With something like this we could get rid of needing to include a second JWT dependency in the mix.

Currently this only supports RSA keys, but could easily be extended if considered a good approach.

anakinj commented on lib/jwt/jwk.rb:
```diff
@@ -0,0 +1,66 @@
+      keypair.private?
+    end
+
+    def public_key
```
Really? Its a Rails fetature....

excpt commented:
Hi @anakinj,

thanks for the PR! This looks very good.

It would be great if you could include an update in the README.md that JWK is supported for RSA keys.

Please ignore the build errors in travis. This seems to be related to the used OpenSSL version in the containers.

anakinj commented:
Hi @excpt.

Thanks for the feedback. I'll do something related to the documentation and was also thinking to add support for EC public keys.

It seems that the Travis build failure is because Ruby 2.2 does not support the openssl gem specified in the gemspec. Maybe we could just drop Ruby 2.2 support (drop the build), the road for ruby 2.2 is ending anyway (https://www.ruby-lang.org/en/news/2018/06/20/support-of-ruby-2-2-has-ended/).

anakinj commented:
Got some time to refactor this, moved the JWK related things into its own module and separated the RSA handling to a completely own class. Also added a code example to the README.md.

Did not add the EC support just yet, i did not quite understand how the parameters should be represented for those keys.

anthonylebrun commented:
Sweet! Thanks @anakinj for working on this. Is this just pending review atm?

anakinj commented:
@anthonylebrun, I'm have nothing pending related to this anymore. Additional algorithms and such can be added later, if there is a need for it.

joelvh commented:
Great to see this merged. When will it be released?

bethesque commented:
Also keen to have this released!

bjeanes commented:
Just wanted to drop in here and thank @anakinj and everyone that helped get this over the line. I just used this in a new integration and found it really straight forward to decode JWTs and validate with JWKs. 🙏 thank you for allowing the JWKs to be loaded from a `.call`-able too!

anakinj commented:
Great to hear that this giving value and thank you for the kind words @bjeanes

excpt commented:
@bethesque, @joelvh 

There is a beta release available. 

Changelog: https://github.com/jwt/ruby-jwt/releases/tag/v2.2.0.pre.beta.0
Rubygem: https://rubygems.org/gems/jwt/versions/2.2.0.pre.beta.0

joelvh commented:
Thanks @excpt !

bethesque commented:
Thanks! We've been using the gem from github source and it's been working well for us. Thanks for your work on this.
